keras_cv/layers/object_detection/non_max_suppression.py

```diff
@@ -0,0 +1,521 @@
+            )
+            sorted_scores = ops.take_along_axis(
+                scores,
+                sorted_scores_indices,
+                axis=1,
+            )
+            sorted_boxes = ops.take_along_axis(
+                boxes,
+                ops.expand_dims(sorted_scores_indices, axis=-1),
+                axis=1,
+            )
+        return sorted_scores, sorted_boxes, sorted_scores_indices
+
+    batch_dims = ops.shape(boxes)[:-2]
+    num_boxes = boxes.shape[-2]
+    boxes = ops.reshape(boxes, [-1, num_boxes, 4])
+    scores = ops.reshape(scores, [-1, num_boxes])
+    batch_size = boxes.shape[0]
+    if score_threshold != float("-inf"):
+        with ops.name_scope("filter_by_score"):
+            score_mask = ops.cast(scores > score_threshold, scores.dtype)
+            scores *= score_mask
+            box_mask = ops.expand_dims(ops.cast(score_mask, boxes.dtype), 2)
+            boxes *= box_mask
+
+    scores, boxes, sorted_indices = _sort_scores_and_boxes(scores, boxes)
+
+    pad = (
+        math.ceil(max(num_boxes, max_output_size) / tile_size) * tile_size
+        - num_boxes
+    )
+    boxes = ops.pad(ops.cast(boxes, "float32"), [[0, 0], [0, pad], [0, 0]])
+    scores = ops.pad(ops.cast(scores, "float32"), [[0, 0], [0, pad]])
+    num_boxes_after_padding = num_boxes + pad
+    num_iterations = num_boxes_after_padding // tile_size
+
+    def _loop_cond(unused_boxes, unused_threshold, output_size, idx):
+        return ops.logical_and(
+            ops.min(output_size) < ops.cast(max_output_size, "int32"),
+            ops.cast(idx, "int32") < num_iterations,
+        )
+
+    def suppression_loop_body(boxes, iou_threshold, output_size, idx):
+        return _suppression_loop_body(
+            boxes, iou_threshold, output_size, idx, tile_size
+        )
+
+    selected_boxes, _, output_size, _ = ops.while_loop(
+        _loop_cond,
+        suppression_loop_body,
+        [
+            boxes,
+            iou_threshold,
+            ops.zeros([batch_size], "int32"),
+            ops.array(0),
+        ],
+    )
+    num_valid = ops.minimum(output_size, max_output_size)
+    idx = num_boxes_after_padding - ops.cast(
+        ops.top_k(
+            ops.cast(ops.any(selected_boxes > 0, [2]), "int32")
+            * ops.cast(
+                ops.expand_dims(ops.arange(num_boxes_after_padding, 0, -1), 0),
+                "int32",
+            ),
+            max_output_size,
+        )[0],
+        "int32",
+    )
+    idx = ops.minimum(idx, num_boxes - 1)
+
+    index_offsets = ops.cast(ops.arange(batch_size) * num_boxes, "int32")
+    take_along_axis_idx = ops.reshape(
+        idx + ops.expand_dims(index_offsets, 1), [-1]
+    )
+
+    # TODO(ianstenbit): Fix bug in tfnp.take_along_axis that causes this hack.
+    # (This will be removed anyway when we use built-in NMS for TF.)
+    if multi_backend() and keras.backend.backend() != "tensorflow":
```

Comment:
ah, I see, this is your approach to the shape issues. looks reasonable.

will still poke around and see if we can fix this at tf level.

Reply:
yeah this is hacky and awful, but it's not going to be around for long because this NMS op won't be used for TF backends so I figured it wasn't worth a lot of investment here.

That said, I think we should invest in fixing the tfnp op and I'm happy to help with that